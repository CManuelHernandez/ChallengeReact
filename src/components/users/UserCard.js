import React from "react";
import Card from "react-bootstrap/Card";
import Button from "react-bootstrap/Button";
import classes from "./Users.module.css";
import PropTypes from "prop-types";

export const UserCard = ({ id, name, email }) => {
  return (
    <Card className={classes.userCardView}>
      <Card.Body>
        <Card.Title>
          <strong>ID :</strong> {id}
        </Card.Title>
        <Card.Text>NAME: {name}</Card.Text>
        <Card.Text style={{ textAlign: "justify" }}>EMAIL: {email}</Card.Text>
        <Button className={classes.btnPrimary} href={`/${id}`}>
          Detail
        </Button>
      </Card.Body>
    </Card>
  );
};

UserCard.protoType = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  email: PropTypes.string.isRequired,
};
