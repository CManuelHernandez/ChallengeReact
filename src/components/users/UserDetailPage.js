import React from "react";
import { Container, Row, Col, Button } from "react-bootstrap";
import Card from "react-bootstrap/Card";
import classes from "./Users.module.css";
import userImg from "../../assets/ninja.jpg";
import { Loading } from "../helpers/Loading";
import { useParams, Navigate, useNavigate } from "react-router-dom";

export const UserDetailPage = () => {
  const { id } = useParams();
  const navigate = useNavigate();

  const [user, setUser] = React.useState([]);
  const [loading, setLoading] = React.useState(false);

  React.useEffect(() => {
    //Api Call of the specific ID
    const getUser = async () => {
      const url = `https://my-user-manager.herokuapp.com/users/${id}`;
      const resp = await fetch(url);
      const results = await resp.json();
      setUser(results);
      setLoading(true);
    };
    getUser();
  }, [id]);

  const handleReturn = () => {
    navigate(-1);
  };

  if (!user) {
    return <Navigate to="/" />;
  }

  return (
    <Container fluid className={classes.userSection}>
      <h2 className={classes.userSubtitle}>User Info</h2>
      <Container className={classes.userWrapper}>
        <Card className={classes.userCardViewDetail}>
          <Row>
            {loading ? ( // Display a Loading when is getting the data from the API
              <>
                <Col md={3} className={classes.userDetailImg}>
                  <Card.Img src={userImg} alt="card-img" />
                </Col>
                <Col md={9} className={classes.userDetailInfo}>
                  <Card.Body
                    className={classes.userDetailInfoText}
                    data-testid={`testUser`}
                  >
                    <Card.Title>ID:{user.id}</Card.Title>
                    <Card.Title>Name: {user.name}</Card.Title>
                    <Card.Title>Email: {user.email}</Card.Title>
                    <Button
                      className={`${classes.btnPrimary} ${classes.btnDetail}`}
                      onClick={handleReturn}
                    >
                      Retrun
                    </Button>
                  </Card.Body>
                </Col>
              </>
            ) : (
              <Loading />
            )}
          </Row>
        </Card>
      </Container>
    </Container>
  );
};
