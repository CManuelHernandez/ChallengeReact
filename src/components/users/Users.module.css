.userSection {
  position: relative;
  padding-top: 150px;
  padding-bottom: 30px;
}
.userCard {
  color: white;
  font-weight: bold;
  padding-top: 50px;
  padding-bottom: 50px;
  padding-left: 25px;
  padding-right: 25px;
  height: auto;
}
.userCardView,
.userCardViewDetail {
  box-shadow: 0 0 15px #b9292f;
  background-color: transparent;
  opacity: 0.9;
  transition: all 0.5s ease 0s;
  height: 100%;
  background-color: #b8bab7;
}
.userCardView:hover,
.userCardViewDetail:hover {
  transform: scale(1.02);
  overflow: hidden;
  box-shadow: 0 0 5px #b9292f;
}
.userCardView {
  background-color: #bbbcbc;
}
.btnPrimary {
  display: flex;
  justify-content: center;
  margin-top: 0.75rem;
  color: #f4e7e6;
  font-weight: bold;
  background-color: #db2529;
  border-color: #db2529;
}
.btnDetail {
  position: absolute;
  right: 0;
  bottom: 0;
  margin: 0 0.75rem 0.75rem 0;
}
.btnPrimary:hover {
  background-color: #b9292f;
  border-color: #b9292f;
}
.btn:focus {
  outline: none;
  box-shadow: none;
}
.userHeading,
.userSubtitle {
  color: whitesmoke;
  font-size: 2.3em;
  font-weight: 500;
  padding-top: 10px;
}
.userSubtitle {
  text-align: center;
  font-size: 1.8em;
}
.userInfoTitle h1,
.userInfoTitle p {
  display: flex;
  justify-content: center;
}
.userDetailInfoText {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: 70%;
}
