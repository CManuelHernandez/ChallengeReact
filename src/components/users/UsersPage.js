import React from "react";
import { Container, Row, Col } from "react-bootstrap";
import { Loading } from "../helpers/Loading";
import { UserCard } from "./UserCard";

import classes from "./Users.module.css";

export const UsersPage = () => {
  const [users, setUsers] = React.useState([]);
  const [loading, setLoading] = React.useState(false);

  React.useEffect(() => {
    getUsers();
  }, []);

  //Api Call
  const getUsers = async () => {
    const url = "https://my-user-manager.herokuapp.com/users";
    const resp = await fetch(url);
    const results = await resp.json();

    const apiUsers = results.map((user) => {
      return {
        id: user.id,
        name: user.name,
        email: user.email,
      };
    });
    setUsers(apiUsers);
    setLoading(true);
  };

  return (
    <Container fluid className={classes.userSection}>
      <Container className={classes.userWrapper}>
        <Col md={12} className={classes.userInfoTitle}>
          <div>
            <h1 className={classes.userHeading}>
              Code Challenge &nbsp;
              <strong className="turquoise"> React </strong>
              &nbsp; Junior
            </h1>
            <h2 className={classes.userSubtitle}>
              Carlos Manuel Hern√°ndez Montero
            </h2>
          </div>
        </Col>

        <Row style={{ justifyContent: "center", paddingBottom: "10px" }}>
          {loading ? ( // Display a Loading when is getting the data from the API
            users.map(
              (
                person // Display Api Data
              ) => (
                <Col
                  md={5}
                  key={person.id}
                  className={classes.userCard}
                  data-testid={`testUser${person.id}`} // Only for Testing Purposes
                >
                  <UserCard
                    md={4}
                    className={classes.userCard}
                    name={person.name}
                    email={person.email}
                    {...person}
                  />
                </Col>
              )
            )
          ) : (
            <Loading />
          )}
        </Row>
      </Container>
    </Container>
  );
};
