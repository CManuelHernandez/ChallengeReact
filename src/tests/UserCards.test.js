import React from "react";
import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";

import { UserCard } from "../components/users/UserCard";

describe("UserCard is render", () => {
  test("should return true", () => {
    render(<UserCard />);
    const title = screen.getByText("ID :");
    expect(title).toBeInTheDocument();
  });
});

describe("Test on <UserCard />", () => {
  test("snapshot testing", () => {
    const tree = renderer.create(<UserCard />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
