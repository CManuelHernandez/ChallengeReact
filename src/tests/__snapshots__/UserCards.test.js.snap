// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <UserCard /> snapshot testing 1`] = `
<div
  className="userCardView card"
>
  <div
    className="card-body"
  >
    <div
      className="card-title h5"
    >
      <strong>
        ID :
      </strong>
       
    </div>
    <p
      className="card-text"
    >
      NAME: 
    </p>
    <p
      className="card-text"
      style={
        Object {
          "textAlign": "justify",
        }
      }
    >
      EMAIL: 
    </p>
    <a
      className="btnPrimary btn btn-primary"
      href="/undefined"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      Detail
    </a>
  </div>
</div>
`;
